session.go

```diff
@@ -991,27 +995,57 @@ func (s *session) loadCommonGlobalVariablesIfNeeded() error {
 	return nil
 }
 
-type txnWithErr struct {
+// txnFuture is a promise, which promises to return a txn in future.
+type txnFuture struct {
```

Comment:
```
s.txn != nil
s.txn.Valid()
s.txnFuture
```

It's quite boring, we can just provide txnFuture and provide method on it...
but maybe we can leave it to later refactor, not in this PR